### ISSUE TITLE
Why not rely on PATH?

### ISSUE THREAD

**Issue by Daniel15** [tokiwinter/cloudns-api#1]
Is there a reason the script specifies absolute paths for every utility it uses?

```
AWK="/usr/bin/awk"
BASENAME="/usr/bin/basename"
CAT="/usr/bin/cat"
CURL="/usr/bin/curl"
CUT="/usr/bin/cut"
DATE="/usr/bin/date"
ECHO="builtin echo"
EVAL="builtin eval"
GETOPTS="builtin getopts"
GREP="/usr/bin/grep"
JQ="/usr/bin/jq"
SED="/usr/bin/sed"
SLEEP="/usr/bin/sleep"
TEST="/usr/bin/test"
TR="/usr/bin/tr"
WC="/usr/bin/wc"
```

This is not very portable. For example, if I compile jq myself, it will normally be in `/usr/local/bin` rather than `/usr/bin`, and Debian usually has `date` in `/bin/date` rather than `/usr/bin/date`.

These are the changes I had to make to get it running on Debian and Ubuntu:
```diff
--- a/bin/cloudns_api.sh
+++ b/bin/cloudns_api.sh
@@ -5,13 +5,13 @@ BASENAME="/usr/bin/basename"
 CAT="/usr/bin/cat"
 CURL="/usr/bin/curl"
 CUT="/usr/bin/cut"
-DATE="/usr/bin/date"
+DATE="/bin/date"
 ECHO="builtin echo"
 EVAL="builtin eval"
 GETOPTS="builtin getopts"
-GREP="/usr/bin/grep"
+GREP="/bin/grep"
 JQ="/usr/bin/jq"
-SED="/usr/bin/sed"
+SED="/bin/sed"
-SLEEP="/usr/bin/sleep"
+SLEEP="/bin/sleep"
 TEST="/usr/bin/test"
 TR="/usr/bin/tr"
```

It would be easier to just rely on these being on the path rather than explicitly specifying the absolute path to them.

**Comment by tokiwinter** [comment]
@Daniel15 - thanks for raising this. This was an artifact of my development, where I wanted to override certain shell `builtin`s, and had some custom compiles of a few utilities that I was switching back and forth between (namely versions of `jq`). I do this during initial development so that I know exactly which binary is being called, or whether a shell `builtin` is being used, and don't rely on `PATH`. For development, I like to be explicit - for example, when I'm testing multiple versions of a utility, I'll often install things to `/usr/local/<tool>-<version>/bin` and then can simply update the appropriate variable to use a different version (I should probably use `stow(1)` instead for this).

That said, the tool is in the wild now, so I'll get this updated to use `PATH` for portability as you suggest. As I've done above, I can use `builtin <blah>` when I want to explicitly call a shell `builtin`. Watch this space. I'll also take a look at those PRs you raised over the next day or two when I get some time.

**Comment by Daniel15** [comment]
> for example, when I'm testing multiple versions of a utility, I'll often install things to /usr/local/<tool>-<version>/bin and then can simply update the appropriate variable to use a different version

You could achieve this by modifying your `PATH` though, right? Running the script like `PATH=/usr/local/<tool>-<version>/bin:$PATH ./cloudns-api.sh` should work 😃 

`stow` is very good for that purpose too. I used to use it to switch between multiple different versions of Mono.

**Comment by tokiwinter** [comment]
True, but that still wouldn't solve the problem of `builtin`s taking precedence. With the variable method, I can also see which version of a particular utility is being used in a given commit. Plus, when there are many tools in play, modifying `PATH` in this way prior to each invocation of the script becomes unwieldy. There are other ways around this though.

I also wanted this to work on Solaris, using a mix of binaries (xpg4, gnu, etc.) and setting up `PATH` there was horrendous. I don't want to support that anymore, so ditching the binary paths in favour of `PATH` is definitely the way to go, and my preferred way of working.

In any case, I'll get the script updated over the next day or two to remove the absolute paths. 

`stow` is a brilliant tool - I use it to manage my dotfiles - I've found it incredibly useful for that too.

**Comment by tokiwinter** [comment]
I haven't forgotten about this - just moved onto other things. I'll fix it, one day.

**Comment by tokiwinter** [comment]
Fixed in #5
